### PR TITLE
[CDAP-20683] Propogate logs from System Worker to Pipeline Logs

### PR DESCRIPTION
**Pull request by dj-smart**
JIRA: [CDAP-20683](https://cdap.atlassian.net/browse/CDAP-20683)

[CDAP-20683]: https://cdap.atlassian.net/browse/CDAP-20683?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

The StartupTask timeout logs have been propagated to pipeline logs, as can be seen in this image: 
image

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeService.java [hunk 1]
```diff
@@ -138,7 +138,7 @@ public final RuntimeInfo run(ProgramDescriptor programDescriptor, ProgramOptions
       } catch (Exception e) {
         controller.failed(e);
         programStateWriter.error(programRunId, e);
-        LOG.error("Exception while trying to run program run {}", programRunId, e);
+        LOG.error("Exception while trying to run program with runID {}", programRunId, e);
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeService.java]: Why do you need this change? program run is a proper term

Comment (dj-smart) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeService.java]: Ok, I misinterpreted term "program run", so I changed it when I came across it. Will revert it back.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java [hunk 2]
```diff
@@ -649,6 +653,7 @@ public TwillController create(@Nullable Callable<Void> startupTask, long timeout
               }
               programStateWriter.error(programRunId, throwable);
             }
+            restoreContext.cancel();
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: Please use try..finally to ensure context is always properly cancelled, even if exception is thrown out. 

BTW: It's a perfect case for try with resources. We just need to either make Cancellable extend AutoClosable or at least setLoggingContext return an AutoClosable result

Comment (dj-smart) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: Refactored the code to use try-with-resources. PTAL.

File: cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java [hunk 3]
```diff
@@ -73,6 +73,17 @@ public void cancel() {
     };
   }
 
+  /**
+   * Allows to restore the logging context.
+   *
+   * @param loggingContext context to set
+   * @return Autocloseable that can be used to revert the logging context and MDC Map to its original
+   */
+  public static AutoCloseable restoreLoggingContext(LoggingContext loggingContext) {
```
Comment (tivv) [inline cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java]: This looks strange. Usually try with resource should return a resource. Any cleanup happens on exit of the block. My recommendation: is to create an interface that implements both Cancelable and AutoClosable and make the setLoggingContext return it

Comment (dj-smart) [reply cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java]: I have made changes based on your suggestion. Let me know if this is in line with the expectation.

File: cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java [hunk 4]
```diff
@@ -59,17 +56,21 @@ public static Cancellable setLoggingContext(LoggingContext context) {
       // MDC will throw this if there is no valid binding. Normally this shouldn't happen, but in case it does,
       // we'll just ignore it as it doesn't affect platform logic at all as we always use loggingContext.
     }
-    return new Cancellable() {
+    return new LoggingContextRestorer() {
       private boolean cancelled;
 
-      @Override
+      @Override // From Cancellable
```
Comment (tivv) [inline cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java:72]: Use JavaDoc as much as possible

File: cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java [hunk 5]
```diff
@@ -59,17 +56,21 @@ public static Cancellable setLoggingContext(LoggingContext context) {
...
       public void cancel() {
         if (Thread.currentThread() == saveCurrentThread && !cancelled) {
           MDC.setContextMap(saveContextMap == null ? Collections.emptyMap() : saveContextMap);
           loggingContext.set(saveLoggingContext);
           cancelled = true;
         }
       }
+      @Override // From AutoCloseable
```
Comment (tivv) [inline cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java:85]: Also here use Javadoc

**Review by tivv — APPROVED** [review]
LGTM with small comments

**Review by masoud-io — APPROVED** [review]

File: cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java (hunk 4 above)
Comment (dj-smart) [reply cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java:72]: Added javadoc

File: cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java (hunk 5 above)
Comment (dj-smart) [reply cdap-common/src/main/java/io/cdap/cdap/common/logging/LoggingContextAccessor.java:85]: Added javadoc

**Pull request merged**